### PR TITLE
add resource and data source of Network ACL.

### PR DESCRIPTION
**Pull request by crab21**
add resource and data source of Network ACL.
***
 $ make testacc TESTARGS=" -run=TestAccDataSourceTencentCloudVpcACL_Basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccDataSourceTencentCloudVpcACL_Basic -timeout 120m
?       github.com/terraform-providers/terraform-provider-tencentcloud  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/gendoc   0.021s [no tests to run]
=== RUN   TestAccDataSourceTencentCloudVpcACL_Basic
--- PASS: TestAccDataSourceTencentCloudVpcACL_Basic (12.47s)
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud     12.495s
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity        [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper     [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit   [no test files]

 $ make testacc TESTARGS=" -run=TestAccTencentCloudVpcAcl"           
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudVpcAcl -timeout 120m
?       github.com/terraform-providers/terraform-provider-tencentcloud  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/gendoc   0.024s [no tests to run]
=== RUN   TestAccTencentCloudVpcAclAttachment_basic
--- PASS: TestAccTencentCloudVpcAclAttachment_basic (25.62s)
=== RUN   TestAccTencentCloudVpcAcl_basic
--- PASS: TestAccTencentCloudVpcAcl_basic (17.66s)
=== RUN   TestAccTencentCloudVpcAclRulesUpdate
--- PASS: TestAccTencentCloudVpcAclRulesUpdate (24.09s)
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud     67.402s
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity        [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper     [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit   [no test files]

### CODE REVIEW

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+				ValidateFunc: validateNotEmpty,
+				Description:  "ID of the VPC instance.",
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateStringLengthInRange(0, 60),
+				Description:  "Name of the network ACL.",
+			},
+			"id": {
```
Comment (Sherlock-Holo) [inline tencentcloud/data_source_tc_vpc_acls.go:47]: z这里要不要叫 acl_id 比较好？

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+		service = VpcService{client: meta.(*TencentCloudClient).apiV3Conn}
+
+		vpcID string
+		name  string
+		id    string
+	)
+
+	if temp, ok := d.GetOk("vpc_id"); ok {
+		tempStr := temp.(string)
+		if tempStr != "" {
```
Comment (Sherlock-Holo) [inline tencentcloud/data_source_tc_vpc_acls.go:207]: 可以不用再判断了

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+			vpcID = tempStr
+		}
+	}
+	if temp, ok := d.GetOk("name"); ok {
+		tempStr := temp.(string)
+		if tempStr != "" {
```
Comment (Sherlock-Holo) [inline tencentcloud/data_source_tc_vpc_acls.go:213]: 同上

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+			name = tempStr
+		}
+	}
+	if temp, ok := d.GetOk("id"); ok {
+		tempStr := temp.(string)
+		if tempStr != "" {
```
Comment (Sherlock-Holo) [inline tencentcloud/data_source_tc_vpc_acls.go:219]: 同上

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+		ingressInfo := info.IngressEntries
+		ingress := make([]map[string]interface{}, 0, len(ingressInfo))
+		for i := range ingressInfo {
+			v := ingressInfo[i]
+			egressMap := make(map[string]interface{}, 5)
+			egressMap["protocol"] = v.Protocol
+			egressMap["port"] = v.Port
+			egressMap["cidr_block"] = v.CidrBlock
+			egressMap["policy"] = v.Action
+			egressMap["description"] = v.Description
```
Comment (Sherlock-Holo) [inline tencentcloud/data_source_tc_vpc_acls.go:262]: 可以直接
```go
egressMap := map[string]interface{}{
    "protocol": v.Protocol,
    "port": v.Port,
}
```

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+			"egress":      egress,
+		}
+		aclList = append(aclList, aclResult)
+		ids = append(ids, *info.NetworkAclId)
+	}
+
+	d.SetId(helper.DataResourceIdsHash(ids))
+	err = d.Set("acl_list", aclList)
+	if err != nil {
+		log.Printf("[CRITAL]%s provider set acl list fail, reason:%v \n ", logId, err)
```
Comment (Sherlock-Holo) [inline tencentcloud/data_source_tc_vpc_acls.go:297]: 结尾不需要 `\n`

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+			},
+			"subnet_ids": {
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: "ID list of the Subnet instance ID.",
+			},
```
Comment (Sherlock-Holo) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:65]: 想一次性绑多个的话，subnet_ids 不应该 forcenew：有一个子网不想要 acl 了，其他子网也得解绑？
想 forcenew 的话，就不要一次性绑定多个了

File: tencentcloud/resource_tc_vpc_acl.go [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+		vpcService = VpcService{client: meta.(*TencentCloudClient).apiV3Conn}
+		vpcID      string
+		name       string
+		ingress    []VpcACLRule
+		egress     []VpcACLRule
+	)
+
+	if temp, ok := d.GetOk("vpc_id"); ok {
+		vpcID = temp.(string)
+	}
```
Comment (Himer) [inline tencentcloud/resource_tc_vpc_acl.go:104]: required属性可以直接下面即可
vpcID = d.Get("vpc_id").(string)
name= d.Get("name").(string)

File: tencentcloud/resource_tc_vpc_acl.go [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = vpcService.AttachRulesToACL(ctx, aclID, ingress, egress)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(aclID)
```
Comment (Himer) [inline tencentcloud/resource_tc_vpc_acl.go:139]: create函数一执行完成后就可以 d.SetId(aclID) ,这样可以防止AttachRulesToACL出错后本地没法追踪生成的aclID

File: tencentcloud/resource_tc_vpc_acl.go [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+		d.SetId("")
+		return nil
+	}
+	if has != 1 {
+		errRet := fmt.Errorf("one acl_id read get %d acl info", has)
+		log.Printf("[CRITAL]%s %s\n", logId, errRet.Error())
+		return errRet
+	}
+
+	_ = d.Set("vpc_id", vpcID)
```
Comment (Himer) [inline tencentcloud/resource_tc_vpc_acl.go:170]: 此刻也开业

File: tencentcloud/resource_tc_vpc_acl.go [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+		service = VpcService{client: meta.(*TencentCloudClient).apiV3Conn}
+		id      = d.Id()
+
+		name          *string
+		ingress       []VpcACLRule
+		egress        []VpcACLRule
+		deleteIngress bool
+		deleteEgress  bool
+	)
+
```
Comment (Himer) [inline tencentcloud/resource_tc_vpc_acl.go:196]: 下面已经修改name了,所以此处可以开始d.Partial(true)

修改name后 d.SetPartial("name")

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+		service   = VpcService{client: meta.(*TencentCloudClient).apiV3Conn}
+		aclId     string
+		subnetIds []string
+		sub_id    string
+	)
+
+	if temp, ok := d.GetOk("acl_id"); ok {
+		aclId = temp.(string)
+		if len(aclId) < 1 {
+			return fmt.Errorf("acl_id should be not empty string")
```
Comment (Himer) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:84]: aclId的check可以放在validatefunc中进行
这样用户plan的时候就可以看到保存
而不是apply阶段出现

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+		subnetIds = helper.InterfacesStrings(temp.([]interface{}))
+	}
+
+	err := service.AssociateAclSubnets(ctx, aclId, subnetIds)
+	if err != nil {
+		return err
+	}
+
+	for _, temp_id := range subnetIds {
+		sub_id = sub_id + "#" + temp_id
```
Comment (Himer) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:97]: strings.join

d.SetId(aclId + "#" + strings.join(subnetIds,"#"))

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+		if e != nil {
+			return retryError(e)
+		}
+		return nil
+	})
+	if err != nil {
+		log.Printf("[CRITAL]%s read acl attachment failed, reason:%s\n", logId, err.Error())
+		return err
+	}
+	time.Sleep(10 * time.Second)
```
Comment (Himer) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:113]: 因为上面已经DescribeByAclId ,可以判断资源已经存在
此处duck不必在sleep

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+	)
+
+	if attachmentId == "" {
+		return fmt.Errorf("attachmentId does not exist")
+	}
+
+	aclId = strings.Split(attachmentId, "#")[0]
+
+	results, err := service.DescribeNetWorkAcls(ctx, aclId, "", "")
+	if err != nil {
```
Comment (Himer) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:137]: 缺少云上已经删除资源后 本地setid("")的操作逻辑

File: tencentcloud/service_tencentcloud_vpc.go [hunk 16]
```diff
@@ -2766,3 +2774,507 @@ func waitEniDetach(ctx context.Context, id string, client *vpc.Client) error {
...
+		request = vpc.NewDisassociateNetworkAclSubnetsRequest()
+		err     error
+
+		subIds []*string
+		aclId  string
+	)
+
+	if attachmentAcl == "" {
+		errRet = fmt.Errorf("DeleteRouteTable can not invoke by empty routeTableId.")
+		return
```
Comment (Himer) [inline tencentcloud/service_tencentcloud_vpc.go:3251]: DisassociateNetworkAclSubnets  can not invoke by empty NetworkAclId

File: tencentcloud/service_tencentcloud_vpc.go [hunk 17]
```diff
@@ -2766,3 +2774,507 @@ func waitEniDetach(ctx context.Context, id string, client *vpc.Client) error {
...
+	}
+
+	for i, v := range strings.Split(attachmentAcl, "#") {
```
Comment (Himer) [inline tencentcloud/service_tencentcloud_vpc.go:3254]: items:=strings.Split(attachmentAcl, "#")
	aclId = items[0]
	subIds = items[1:]

File: tencentcloud/service_tencentcloud_vpc.go [hunk 18]
```diff
@@ -2766,3 +2774,507 @@ func waitEniDetach(ctx context.Context, id string, client *vpc.Client) error {
...
+		if i == 0 {
+			aclId = v
+		}
```
Comment (Himer) [inline tencentcloud/service_tencentcloud_vpc.go:3227]: aclId = strings.Split(attachmentAcl, "#")[0]

File: tencentcloud/resource_tc_vpc_acl.go (hunk 10 above)
Comment (oliverpei) [inline tencentcloud/resource_tc_vpc_acl.go:170]: 其他参数也需要read

File: tencentcloud/resource_tc_vpc_acl.go (hunk 11 above)
Comment (oliverpei) [inline tencentcloud/resource_tc_vpc_acl.go:196]: d.Partial(true) 写到这里

File: tencentcloud/resource_tc_vpc_acl.go [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+	if d.HasChange("name") {
+		name = helper.String(d.Get("name").(string))
+		err := service.ModifyVpcNetworkAcl(ctx, &id, name)
+		if err != nil {
+			return nil
+		}
```
Comment (oliverpei) [inline tencentcloud/resource_tc_vpc_acl.go:202]: name也需要d.SetPartial('name")

File: tencentcloud/data_source_tc_vpc_acls.go (hunk 1 above)
Comment (crab21) [reply tencentcloud/data_source_tc_vpc_acls.go:47]: > 这里要不要叫 acl_id 比较好？

这个参考了其它的，如VPC等等，id可以满足需求。

File: tencentcloud/data_source_tc_vpc_acls.go (hunk 2 above)
Comment (crab21) [reply tencentcloud/data_source_tc_vpc_acls.go:207]: > 可以不用再判断了

如果是“”字符串，就没有必要给赋值，再后续进行查询了

File: tencentcloud/resource_tc_vpc_acl.go [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+
+	if deleteEgress {
+		if err := service.DeleteACLRulesByChoose(ctx, id, nil, egress, DeleteEgress); err != nil {
+			return err
+		}
+
+		d.SetPartial("egress")
+
+		egress = nil
+	}
```
Comment (oliverpei) [inline tencentcloud/resource_tc_vpc_acl.go:284]: 修改的逻辑跟安全组不一样，这里逻辑需要再梳理下

File: tencentcloud/data_source_tc_vpc_acls.go (hunk 5 above)
Comment (crab21) [reply tencentcloud/data_source_tc_vpc_acls.go]: > 可以直接
> 
> ```go
> egressMap := map[string]interface{}{
>     "protocol": v.Protocol,
>     "port": v.Port,
> }
> ```

fixed

File: tencentcloud/data_source_tc_vpc_acls.go (hunk 6 above)
Comment (crab21) [reply tencentcloud/data_source_tc_vpc_acls.go:297]: > 结尾不需要 `\n`

fixed

File: tencentcloud/resource_tc_vpc_acl_attachment.go (hunk 7 above)
Comment (crab21) [reply tencentcloud/resource_tc_vpc_acl_attachment.go]: > 想一次性绑多个的话，subnet_ids 不应该 forcenew：有一个子网不想要 acl 了，其他子网也得解绑？
> 想 forcenew 的话，就不要一次性绑定多个了

fixed

File: tencentcloud/resource_tc_vpc_acl.go (hunk 8 above)
Comment (crab21) [reply tencentcloud/resource_tc_vpc_acl.go]: > required属性可以直接下面即可
> vpcID = d.Get("vpc_id").(string)
> name= d.Get("name").(string)

fixed

File: tencentcloud/resource_tc_vpc_acl.go (hunk 9 above)
Comment (crab21) [reply tencentcloud/resource_tc_vpc_acl.go]: > create函数一执行完成后就可以 d.SetId(aclID) ,这样可以防止AttachRulesToACL出错后本地没法追踪生成的aclID

fixed

File: tencentcloud/resource_tc_vpc_acl.go (hunk 10 above)
Comment (crab21) [reply tencentcloud/resource_tc_vpc_acl.go]: > 此处也可以 set一下名字
> 如果ingress&egress也可以读取出来set一下就更完美了

fixed

File: tencentcloud/resource_tc_vpc_acl.go (hunk 11 above)
Comment (crab21) [reply tencentcloud/resource_tc_vpc_acl.go:207]: > 下面已经修改name了,所以此处可以开始d.Partial(true)
> 
> 修改name后 d.SetPartial("name")

fixed

File: tencentcloud/resource_tc_vpc_acl.go (hunk 20 above)
Comment (crab21) [reply tencentcloud/resource_tc_vpc_acl.go]: > 修改的逻辑跟安全组不一样，这里逻辑需要再梳理下

fixed

**Comment by crab21** [comment]
$ make testacc TESTARGS=" -run=TestAccTencentCloudVpcAcl"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudVpcAcl -timeout 120m
?       github.com/terraform-providers/terraform-provider-tencentcloud  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/gendoc   (cached) [no tests to run]
=== RUN   TestAccTencentCloudVpcAclAttachment_basic
--- PASS: TestAccTencentCloudVpcAclAttachment_basic (6.49s)
=== RUN   TestAccTencentCloudVpcAcl_basic
--- PASS: TestAccTencentCloudVpcAcl_basic (18.11s)
=== RUN   TestAccTencentCloudVpcAclRulesUpdate
--- PASS: TestAccTencentCloudVpcAclRulesUpdate (44.87s)
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud     69.486s
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity        [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper     [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit   [no test files]
***
 $ make testacc TESTARGS=" -run=TestAccDataSourceTencentCloudVpcACL_Basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccDataSourceTencentCloudVpcACL_Basic -timeout 120m
?       github.com/terraform-providers/terraform-provider-tencentcloud  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/gendoc   (cached) [no tests to run]
=== RUN   TestAccDataSourceTencentCloudVpcACL_Basic
--- PASS: TestAccDataSourceTencentCloudVpcACL_Basic (12.40s)
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud     12.432s
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity        [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/internal/helper     [no test files]
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/ratelimit   [no test files]

File: examples/tencentcloud-vpc/main.tf [hunk 21]
```diff
@@ -11,3 +11,21 @@ data "tencentcloud_vpc_instances" "tags_instances" {
...
+
+resource "tencentcloud_vpc_acl" "foo" {
+  vpc_id  = data.tencentcloud_vpc_instances.default.instance_list.0.vpc_id
+  name    = "test_acl_update"
+  ingress = ["ACCEPT#192.168.1.0/24#800#TCP", "ACCEPT#192.168.1.0/24#800-900#TCP",]
+  egress  = ["ACCEPT#192.168.1.0/24#800#TCP", "ACCEPT#192.168.1.0/24#800-900#TCP",]
+}
+
+resource "tencentcloud_vpc_acl_attachment" "attachment" {
+  acl_id     = tencentcloud_vpc_acl.foo.id
```
Comment (gailwang) [inline examples/tencentcloud-vpc/main.tf:25]: 后面的取名如default，foo attachment可以统一一下，叫example

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 22]
```diff
@@ -0,0 +1,289 @@
...
+						"ingress": {
+							Type:        schema.TypeList,
+							Computed:    true,
+							Description: "Inbound rules of the network ACL.",
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"protocol": {
+										Type:        schema.TypeString,
+										Computed:    true,
+										Description: "Type of ip protocol.",
```
Comment (gailwang) [inline tencentcloud/data_source_tc_vpc_acls.go:127]: IP统一一下格式

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 23]
```diff
@@ -0,0 +1,289 @@
...
+									"port": {
+										Type:        schema.TypeString,
+										Computed:    true,
+										Description: "Range of the port.",
+									},
+									"policy": {
+										Type:        schema.TypeString,
+										Computed:    true,
+										Description: "Rule policy of.",
+									},
```
Comment (gailwang) [inline tencentcloud/data_source_tc_vpc_acls.go:138]: of啥？

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 24]
```diff
@@ -0,0 +1,289 @@
...
+									"port": {
+										Type:        schema.TypeString,
+										Computed:    true,
+										Description: "Range of the port.",
+									},
+									"policy": {
+										Type:        schema.TypeString,
+										Computed:    true,
+										Description: "Rule policy of.",
```
Comment (gailwang) [inline tencentcloud/data_source_tc_vpc_acls.go:171]: ？

File: tencentcloud/data_source_tc_vpc_acls.go [hunk 25]
```diff
@@ -0,0 +1,289 @@
...
+	if err != nil {
+		return err
+	}
+
+	aclList := make([]map[string]interface{}, 0, len(networkAcls))
+	ids := make([]string, 0, len(networkAcls))
+
+	for _, info := range networkAcls {
+		subnetInfo := info.SubnetSet
+		subnets := make([]map[string]interface{}, 0, len(subnetInfo))
```
Comment (gailwang) [inline tencentcloud/data_source_tc_vpc_acls.go:215]: 这个地方用:=只能保存最后一条subnetInfo

File: tencentcloud/resource_tc_vpc_acl.go [hunk 26]
```diff
@@ -0,0 +1,306 @@
...
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "Ingress rules. A rule must match the following format: [action]#[cidr_ip]#[port]#[protocol]. The available value of 'action' is `ACCEPT` and `DROP`. The 'cidr_ip' must be an IP address network or segment. The 'port' valid format is `80`, `80,443`, `80-90` or `ALL`. The available value of 'protocol' is `TCP`, `UDP`, `ICMP` and `ALL`. When 'protocol' is `ICMP` or `ALL`, the 'port' must be `ALL`.",
+			},
+			"egress": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "Egress rules. A rule must match the following format: [action]#[cidr_ip]#[port]#[protocol]. The available value of 'action' is `ACCEPT` and `DROP`. The 'cidr_ip' must be an IP address network or segment. The 'port' valid format is `80`, `80,443`, `80-90` or `ALL`. The available value of 'protocol' is `TCP`, `UDP`, `ICMP` and `ALL`. When 'protocol' is `ICMP` or `ALL`, the 'port' must be `ALL`.",
+			},
+
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl.go:79]: 去掉空行

File: tencentcloud/resource_tc_vpc_acl.go [hunk 27]
```diff
@@ -0,0 +1,306 @@
...
+	for i := range info.EgressEntries {
+		if info.EgressEntries[i].Port == nil || *info.EgressEntries[i].Port == "" {
+			continue
+		}
+		result := strings.Join([]string{
+			*info.EgressEntries[i].Action,
+			*info.EgressEntries[i].CidrBlock,
+			*info.EgressEntries[i].Port,
+			*info.EgressEntries[i].Protocol,
+		}, "#")
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl.go:177]: 使用FILE_SP常量

File: tencentcloud/resource_tc_vpc_acl.go [hunk 28]
```diff
@@ -0,0 +1,306 @@
...
+	for i := range info.IngressEntries {
+		if info.IngressEntries[i].Port == nil || *info.IngressEntries[i].Port == "" {
+			continue
+		}
+		result := strings.Join([]string{
+			*info.IngressEntries[i].Action,
+			*info.IngressEntries[i].CidrBlock,
+			*info.IngressEntries[i].Port,
+			*info.IngressEntries[i].Protocol,
+		}, "#")
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl.go:191]: 同上

File: tencentcloud/resource_tc_vpc_acl.go [hunk 29]
```diff
@@ -0,0 +1,306 @@
...
+			newIngress := helper.InterfacesStrings(new.([]interface{}))
+			for _, egressStr := range newIngress {
+				liteRule, err := parseACLRule(egressStr)
+				if err != nil {
+					return err
+				}
+				egress = append(egress, liteRule)
+			}
+		}
+	}
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl.go:270]: 这里的ingress 和egress不需要d.SetPartial么？

File: tencentcloud/resource_tc_vpc_acl.go [hunk 30]
```diff
@@ -0,0 +1,306 @@
...
+
+	if err := service.ModifyNetWorkAclRules(ctx, id, ingress, egress); err != nil {
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl.go:272]: 是否判断d.HasChange("egress")||d.HasChange("ingress")再去调用Modify函数？

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 31]
```diff
@@ -0,0 +1,137 @@
...
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"acl_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateNotEmpty,
+				Description:  "Id of the attached ACL.",
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:53]: Id ID 统一下

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+	subnetId := d.Get("subnet_id").(string)
+
+	subnetIds = append(subnetIds, subnetId)
+
+	err := service.AssociateAclSubnets(ctx, aclId, subnetIds)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(aclId + "#" + subnetId)
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:84]: "#" --FILED_SP

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 33]
```diff
@@ -0,0 +1,137 @@
...
+		logId        = getLogId(contextNil)
+		ctx          = context.WithValue(context.TODO(), logIdKey, logId)
+		service      = VpcService{client: meta.(*TencentCloudClient).apiV3Conn}
+		attachmentId = d.Id()
+		aclId        string
+	)
+
+	if attachmentId == "" {
+		return fmt.Errorf("attachmentId does not exist")
+	}
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:103]: 讲道理不需要这么判断

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 34]
```diff
@@ -0,0 +1,137 @@
...
+
+	aclId = strings.Split(attachmentId, "#")[0]
+
+	results, err := service.DescribeNetWorkAcls(ctx, aclId, "", "")
+	if err != nil {
+		return err
+	}
+	if len(results) > 0 && len(results[0].SubnetSet) < 1 {
+		d.SetId("")
+		return fmt.Errorf("[TECENT_TERRAFORM_CHECK][ACL attachment][Read] check: acl attachment  is not create")
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:113]: 这个错误为啥是not create？ 如果Describe不存在 应该是is not exist  not create只能在create函数内判断吧？

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 35]
```diff
@@ -0,0 +1,137 @@
...
+		service      = VpcService{client: meta.(*TencentCloudClient).apiV3Conn}
+		attachmentId = d.Id()
+		aclId        string
+	)
+
+	if attachmentId == "" {
+		return fmt.Errorf("attachmentId does not exist")
+	}
+
+	aclId = strings.Split(attachmentId, "#")[0]
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:105]: FILED_SP

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 36]
```diff
@@ -0,0 +1,137 @@
...
+	var (
+		logId         = getLogId(contextNil)
+		ctx           = context.WithValue(context.TODO(), logIdKey, logId)
+		service       = VpcService{client: meta.(*TencentCloudClient).apiV3Conn}
+		attachmentAcl = d.Id()
+	)
+
+	err := service.DeleteAclAttachment(ctx, attachmentAcl)
+	if err != nil {
+		log.Printf("[CRITAL]%s delete acl attachment failed, reason:%s\n", logId, err.Error())
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:130]: acl ACL 统一一个

File: tencentcloud/resource_tc_vpc_acl_attachment.go [hunk 37]
```diff
@@ -0,0 +1,137 @@
...
+		return err
+	}
+	d.SetId("")
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_attachment.go:133]: 不需要这个，再次去Read的时候，Read函数内有这个逻辑：没有资源会自动设置SetId("")

File: tencentcloud/resource_tc_vpc_acl_attachment_test.go [hunk 38]
```diff
@@ -0,0 +1,87 @@
...
+		service := VpcService{client: testAccProvider.Meta().(*TencentCloudClient).apiV3Conn}
+
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("[TECENT_TERRAFORM_CHECK][ACL attachment][Exists] check:  %s is not found", n)
+		}
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("[TECENT_TERRAFORM_CHECK][ACL attachment][Exists] check: id is not set")
+		}
+		err := service.DescribeByAclId(ctx, rs.Primary.ID)
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_attachment_test.go:59]: 这里最好统一成这个_, has, err:=的写法，有些代码比较陈旧不要参考，得到has为true则检查失败。

File: tencentcloud/resource_tc_vpc_acl_test.go [hunk 39]
```diff
@@ -0,0 +1,214 @@
...
+func testAccCheckVpcACLDestroy(s *terraform.State) error {
+	logId := getLogId(contextNil)
+	ctx := context.WithValue(context.TODO(), logIdKey, logId)
+
+	service := VpcService{client: testAccProvider.Meta().(*TencentCloudClient).apiV3Conn}
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "tencentcloud_vpc_acl" {
+			continue
+		}
+		time.Sleep(5 * time.Second)
```
Comment (gailwang) [inline tencentcloud/resource_tc_vpc_acl_test.go:122]: 这里为啥要加sleep？

Comment (crab21) [reply tencentcloud/resource_tc_vpc_acl_test.go:122]: > 这里为啥要加sleep？

outdated

Comment (crab21) [reply tencentcloud/resource_tc_vpc_acl_test.go]: > 这里为啥要加sleep？

all fixed

**Pull request merged**